src/app/nodetool/qgsnodetool.cpp

```diff
@@ -418,7 +418,7 @@ void QgsNodeTool::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
     QList<Vertex> nodes;
 
     // for each editable layer, select nodes
-    Q_FOREACH ( QgsMapLayer *layer, canvas()->layers() )
+    for ( QgsMapLayer *layer : canvas()->layers() )
```

Comment:
Does that not force a detach?

Reply:
Obviously not: I have used qgsAsConst there before, but it failed to compile because of this in qgis.h:
```
template <typename T>
void qgsAsConst( const T && ) = delete;
```
So my assumption is that the container is actually being moved, not copied, so there is no detach happening.

Reply:
From what I've read, qgsAsConst is based on qAsConst which is based on std::as_const which is - for whatever reason - explicitly only compatible with lvalues and not with rvalues.

The general recommendation (see https://www.kdab.com/goodbye-q_foreach/) is

```
const auto strings = functionReturningQStringList();
for (const QString &s : strings)
    doSomethingWith(s);
```

Reply:
Hmm in this case however the returned list from layers() is built on demand so there should not be any detach anyway, right?

Reply:
@m-kuhn that's my understanding too. The reason for the rvalue deletion is covered here: https://stackoverflow.com/questions/39051460/why-does-as-const-forbid-rvalue-arguments

Reply:
@wonder-sk I believe the use of the range based for FORCES a deep copy, even though it normally wouldn't cause a detach in this case.

Reply:
Okay - here we go with another update... noew I am not sure if I really prefer standard range for loop compared to non-standard Q_FOREACH that is easier to use :-)

Reply:
Me either, but if it's being removed from Qt we don't really have a choice here...

Reply:
Same here, KDAB had to take a lot of criticism for deprecating Q_FOREACH...

Something that has not been widely discussed but that I can imagine to be problematic is unconsting methods.

When removing const in the following situation it's just way too easy to accidentally forget to update the loop

```
void doSomething() const
{
    for ( const &QString s, mStrings )
        ...
}
```

or changing

```
const QStringList &getNames() const;
```

to

```
QStringList getNames() const;
```

Reply:
> Me either, but if it's being removed from Qt we don't really have a choice here...

Latest discussions are "it's deprecated but not going to be removed" from what I could see, so we still have the choice.

Q_FOREACH may perform a tiny bit worse than a correctly used range based for loop but at the same time it's a thousand times safer to not accidentally introduce a detach, plus it's so easy to use.

Reply:
I still think it's worthwhile to start switching. Qt upstream has a history of changing decisions like this, and if it's marked deprecated it's still upstream telling us "you shouldn't be using this". 

Fwiw, clazy has a tool which auto detects detachment caused by range based for loops. I plan on semi-regularly running this over the code base.